### PR TITLE
New Module: Add github_deploy_key module (source_control/github_deploy_key)

### PR DESCRIPTION
**Pull request by bincyber**
##### SUMMARY

Added a new module for managing deploy keys for GitHub repositories.

Currently, this can be accomplished using the uri module, however it requires some effort to achieve idempotency. It's better to have a specific module that can do the same thing in 1 task.

Additionally, this module supports authenticating to GitHub's API using username and password, username and password and 2-factor authentication code (OTP), OAuth2 token, or personal access token.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

github_deploy_key

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /vagrant/ansible.cfg
  configured module search path = [u'modules/']
  python version = 2.7.5 (default, Nov 20 2015, 02:00:19) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### ADDITIONAL INFORMATION

A module for managing deploy keys for GitHub repositories was initially requested in ansible/ansible-modules-core#2626. I began a discussion on the [ansible-devel](https://groups.google.com/forum/#!msg/ansible-devel/XY9XUYFRZvQ/yqOhynX1EQAJ;context-place=forum/ansible-devel) Google Group, but nobody was interested in developing this module. So, I finally got around to doing it myself.

### CODE REVIEW

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py:2]: Add GNU GPL like [github_issue](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/source_control/github_issue.py#L4-L17)

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Remove spaces before operator `=`

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: This condition can come early

```
if username is not None and password is not None:
    self.auth = (self.username, self.password)
    if self.otp is not None:
        self.headers = {"X-GitHub-OTP": self.otp}
```

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: I would prefer `format` here.

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: This can be simplified as `if resp.json()`

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Print exception using `get_exception()`, it will easy for debugging purpose.

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Remove spaces before operator `=`

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Remove spaces before operator.

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: use `fetch_url` instead of `requests.get`

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_deploy_key.py]: version_added: "2.4"

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Please add `no_log=True`

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Please add `no_log=True`

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_deploy_key.py]: What would be needed to support checkmode?h

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@pcgentry @adrianmoisey I believe you two are maintainers for other GitHub modules, it would be good to get your input on this PR. Thanks in advance.

**Comment by bincyber** [comment]
Thanks @Akasurde and @gundalow for your feedback on this PR. I believe I've addressed all of your comments in my latest commit.

File: lib/ansible/modules/source_control/github_deploy_key.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_deploy_key.py]: Line too long.

**Comment by Akasurde** [comment]
@bincyber Shippable is not happy with changes. Can you please fix [issues](https://app.shippable.com/github/ansible/ansible/runs/21883/1/tests) to match coding style?

**Comment by bincyber** [comment]
@Akasurde I've corrected the PEP8 issues and rebased this PR with devel per https://docs.ansible.com/ansible/dev_guide/developing_rebasing.html

**Review by Akasurde — APPROVED** [review]
LGTM

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Looking good, would like to get it review by some other people

@pcgentry @adrianmoisey I believe you two are maintainers for other GitHub modules, it would be good to get your input on this PR. Thanks in advance.

**Comment by gundalow** [comment]
@bincyber Poke me in a week if and if there isn't any other feedback I'll merge it

**Comment by bincyber** [comment]
Thanks @Akasurde @gundalow !

**Comment by bincyber** [comment]
@gundalow Can we get this PR merged please?

**Comment by gundalow** [comment]
Triggering ci

**Comment by gundalow** [comment]
Merged into `devel`, so this will be released in Ansible 2.4

**Comment by bincyber** [comment]
Thanks @gundalow, appreciate it

**Pull request merged**
